### ISSUE TITLE
healthcheck shows neoclip/X11  on wayland

### ISSUE THREAD

**Issue by mskyaxl** [matveyt/neoclip#5]
Hi, I'm running PopOS with wayland enabled.

```
➜  plugin git:(master) ✗ lsb_release -a
No LSB modules are available.
Distributor ID: Pop
Description:    Pop!_OS 22.04 LTS
Release:        22.04
Codename:       jammy
```


I installed neoclip following the instruction from `:h neoclip-build` but when running `checkhealth` I see that neoclip/x11 is shown and getting access to the system clipboard does not work...

```
──────────────────                                                                      
provider: health#provider#check                                                                                                                     
                                                                                                                                                    
Clipboard (optional)                                                                                                                                
- OK Clipboard tool found: neoclip/X11  
```

My expectation is that I should see something like `neoclip/wayland'.

I saw that the help mentions that wayland should support `wlr-data-control` but I don't know how to check if my setup has support of not. But shouldn't `checkhealth` display some error id there is no support instead of falling back to `neoclip/x11`?

Do you have any idea on how I can make it work?


My nvim version is:

```
nvim --version
NVIM v0.9.0-dev
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wno-unused-result -Wimplicit-fallthrough -Wvla -fno-common -fdiagnostics-color=auto -fstack-protector-strong -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DMIN_LOG_LEVEL=3 -DNVIM_UNIBI_HAS_VAR_FROM -I/usr/include/luajit-2.1 -I/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/build/src/nvim/auto -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/build/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/build/cmake.config -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/src -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include -I/build/neovim-FQlxiH/neovim-0.9.0~ubuntu1+git202302191603-bfe6b4944-333b5866f/.deps/usr/include

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```




My neovim config is here; https://github.com/mskyaxl/dotfiles/blob/090001086d8548df65f2c2ffe8224cffdcc5488c/neovim/init.lua#L34

**Comment by matveyt** [comment]
Hi, the protocol support depends on server (aka composer) implementation, not on OS distro. And the server is usually one bundled with the (active) desktop. As I guess, you have Gnome (with Mutter server), and so it is expected not to work.

Considering custom health check, well, I haven't implemented it yet. Currently this is just nvim builtin check that only prints the clipboard provider id. But, maybe, I will add one in the future.

**Comment by mskyaxl** [comment]
Thanks for the explanation. BTW which do you know to be compatible and how can I check?(Should this be in the output of wayland-info for example?)

If this does not work under Gnome maybe it is worth mentioning in the readme. It would save time for others so they don't try it...

**Comment by matveyt** [comment]
1. Neoclip is compatible with Gnome. It simply uses X11 API instead of Wayland. But this is not to be considered as a limitation for user, as Gnome always keeps both clipboards in sync.
2. I don't exactly remember what utilities can show this info. But, well, any one capable of printing out `wl_registry` contents (subscribing to `wl_registry` is practically the first thing to do for any Wayland app after connecting to the display, as this is the only place where it can find out what the composer can do at all). There's even complete example code given at https://wayland-book.com/registry/binding.html
3. KDE/KWin is tested and works with Neoclip/Wayland. Also, "wlr" interfaces come from wlroots project, so I expect it to work with Sway too. Not for Gnome. Not for Weston.
4. There was mentioning of this "Gnome issue" in the docs. But somehow I deleted it in the last edits. Maybe, I will revise it once again.

**Comment by mskyaxl** [comment]
1. It does work on Gnome but only when Wayland is not used.:)
2,3,4. Thanks!

**Comment by matveyt** [comment]
That's strange. Is that for the plugin only, or the two clipboards are out of sync for the whole system? (Btw. Most of apps nowadays are toolkit-based, so it is possible to select backend individually by setting environment variables such as QT_QPA_PLATFORM or GDK_BACKEND, or by cmdline options, such as `-platform xxx` and so on).

**Comment by mskyaxl** [comment]
I'm not sure what is the problem. I'm still investigating... It is hard to see because most of my tools are running under wayland. Nevertheless the behavior is inconsistent - sometimes it does sync, sometimes it does not.

It appears that when xclipboard is running the 2 clipboards are always synchronized - copying from nvim to gedit worked - but I find it annoying that I have to launch xclipboard to get it working.

**Comment by mskyaxl** [comment]
TLDR:
I made it work with some workarounds: install https://github.com/dnut/clipboard-sync and deamonise it with systemd.

Longer explanation:
Wayland is not enabled by default in Pop!_OS(with the default DE) and it is still experimental. The DE dose not have `wlr-data-control` support thus neoclip will fallback to X11. At least for Pop!_OS 22.04, the X11 clipboard and wayland is not synchronized except for the situation in which a X11 clipboard manager is running. I played a lot with a lot of clipboard managers(Gpaste, CopyQ, Clipboard Indicator, diodon) and with copying pasting from X11 windows to wayland windows and for me the clipboards where syncronised only when xclipboard or clipboard-sync where running. Since clipboard-sync can be daemonised and runs without a GUI I will go with it.

Ironically clipboard-sync also requires `wlr-data-control` and fallbacks to X11Clipboard. In the end I don't know who really does the synchronization but it works...

```
feb 26 11:52:51 pop-os-framework env[39677]: 2023-02-26 11:52:51 -  INFO - starting clipboard sync
feb 26 11:52:51 pop-os-framework env[39677]: 2023-02-26 11:52:51 - ERROR - wayland-0 does not support zwlr_data_control_manager_v1, trying with X11Clipboard...
feb 26 11:52:51 pop-os-framework env[39677]: 2023-02-26 11:52:51 -  INFO - Using clipboards: [X11Clipboard { display: ":0" }]
```

@matveyt, I'm not sure how valuable is this info for anyone else facing the same issue. Maybe it is worth adding a section with "troubleshooting" ?

**Comment by matveyt** [comment]
Let's recall how clipboards really work from the beginning,

1. In Windows and MacOS clipboard contents resides in system memory and is owned by OS core. On the other hand, the *nix core doesn't know anything about clipboards. This implies that,
    1a. Clipboard only exists if there's some GUI server running (that actually provides service);
    1b. Every GUI server and session has a clipboard of its own;
    1c. An application must be linked against GUI library (or a toolkit) in order to access clipboard;

2. Both X11 and Wayland act as pure "arbiter" providing some sort of inter-process communication. The clipboard data is always owned by some application and is transferred to another by request (cf. client-server model). This implies that,
    2a. Before an application can access clipboard it must create GUI window (aka surface) and be able to process various windows messages (aka events); this is true even if only reading the clipboard, as all IPC goes asynchronously;
    Workaround: non-GUI application can create a hidden window;
    2b. If clipboard owner application exits then clipboard contents gets lost;
    Workaround: "clipboard manager" is an application that always resides in memory and re-gains clipboard ownership every time the contents changes;

3. Wayland sets an extra limitation: only focused window may access clipboard; there's no way to send request, the server only informs application after it gets the focus. This implies that,
    3a. If any non-GUI app wants to access clipboard it has to use extended protocol, such as `zwlr_data_control_manager_v1`;
    3b. Or it may grab the focus forcefully (which, obviously, makes window flickering, this is how `wl-clipboard` utilities work when no extended protocol is available);

>At least for Pop!_OS 22.04, the X11 clipboard and wayland is not synchronized except for the situation in which a X11 clipboard manager is running. I played a lot with a lot of clipboard managers(Gpaste, CopyQ, Clipboard Indicator, diodon) and with copying pasting from X11 windows to wayland windows and for me the clipboards where syncronised only when xclipboard or clipboard-sync where running.

Every session (including a nested one) has its own clipboard (item 1b). The autosync should work when both servers really run within shared session (e.g. Wayland with XWayland compatibility layer; or Windows+WSL+Wayland+X11 three-in-one, etc). Or if you have some special "synchronizer" application.

>Ironically clipboard-sync also requires wlr-data-control

This is quite expected, see item 3 and 3a.

>I'm not sure how valuable is this info for anyone else facing the same issue. Maybe it is worth adding a section with "troubleshooting" ?

I still consider rewriting the docs.

BTW. Nvim is console application but nvim-qt is GUI. So if nvim runs inside nvim-qt (or such) then it's also possible to access clipboard by routing requests to GUI app. However, request format differs between nvim GUIs, and it won't work at all for terminal application, obviously.

**Comment by matveyt** [comment]
Fixed by 9a66d84.
